numpy/core/tests/test_multiarray.py

```diff
@@ -7814,6 +7814,206 @@ def test_error_if_stored_buffer_info_is_corrupted(self, obj):
+
+                    res = np.array(arr,
+                                   copy=np._CopyMode.NEVER,
+                                   dtype=int2)
+                    assert res is arr or res.base is arr
+
+                else:
+                    # Casting is necessary, assert copy works:
+                    for copy in self.false_vals:
+                        res = np.array(arr, copy=copy, dtype=int2)
+                        assert res is not arr and res.flags.owndata
+                        assert_array_equal(res, arr)
+
+                    assert_raises(ValueError, np.array,
+                                  arr, copy=np._CopyMode.NEVER,
+                                  dtype=int2)
+                    assert_raises(ValueError, np.array,
+                                  arr, copy=None,
+                                  dtype=int2)
+
+    def test_buffer_interface(self):
+
+        # Buffer interface gives direct memory access (no copy)
+        arr = np.arange(10)
+        view = memoryview(arr)
+
+        # Checking bases is a bit tricky since numpy creates another
+        # memoryview, so use may_share_memory.
+        for copy in self.true_vals:
+            res = np.array(view, copy=copy)
+            assert not np.may_share_memory(arr, res)
+        for copy in self.false_vals:
+            res = np.array(view, copy=copy)
+            assert np.may_share_memory(arr, res)
+        res = np.array(view, copy=np._CopyMode.NEVER)
+        assert np.may_share_memory(arr, res)
+
+    def test_array_interfaces(self):
+        # Array interface gives direct memory access (much like a memoryview)
+        base_arr = np.arange(10)
+
+        class ArrayLike:
+            __array_interface__ = base_arr.__array_interface__
+
+        arr = ArrayLike()
+
+        for copy, val in [(True, None), (np._CopyMode.ALWAYS, None),
+                          (False, arr), (np._CopyMode.IF_NEEDED, arr),
+                          (np._CopyMode.NEVER, arr)]:
+            res = np.array(arr, copy=copy)
+            assert res.base is val
+
+    def test___array__(self):
+        base_arr = np.arange(10)
+
+        class ArrayLike:
+            def __array__(self):
+                # __array__ should return a copy, numpy cannot know this
+                # however.
+                return base_arr
+
+        arr = ArrayLike()
+
+        for copy in self.true_vals:
+            res = np.array(arr, copy=copy)
+            assert_array_equal(res, base_arr)
+            # An additional copy is currently forced by numpy in this case,
+            # you could argue, numpy does not trust the ArrayLike. This
+            # may be open for change:
+            assert res is not base_arr
+
+        for copy in self.false_vals:
+            res = np.array(arr, copy=False)
+            assert_array_equal(res, base_arr)
+            assert res is base_arr  # numpy trusts the ArrayLike
+
+        assert np.array(arr, copy=np._CopyMode.NEVER) is base_arr
```

Comment:
```suggestion
        with pytest.raises(TypeError):
            np.array(arr, copy=np._CopyMode.NEVER)
```
@czgdp1807 these tests look good, but my point is and was that this has to fail. `__array__()` probably should never get called in this case (because that may already create that copy we explicitly want to avoid).  The older documentation even suggested it should return a copy IIRC, which somewhat makes sense, if you want a view, you could probably use `__array_interface__` (the current documentation of `__array__` seems to just tell you what not to use it for).

This can be done, but you will have to pass that information deeper into the `DTypeAndShape` Discovery.

Reply:
(Or I guess `ValueError`, is what the rest currently gives.  That seems right?)

Note that it still is not rejected, a quick copy-pastable example:
```python
class m:
    def __array__(self):
        return np.arange(1000)
np.array(m(), copy=np._CopyMode.NEVER, dtype=np.float64)
```
which _should_ raise.

Reply:
With the following diff, I am able to obtain the following error with the above example,

**Error**

```python
Traceback (most recent call last):
  File "/Users/czgdp1807/numpy_project/debug.py", line 7, in <module>
    np.array(m(), copy=np._CopyMode.NEVER, dtype=np.float64)
ValueError: Calling __array__ in never copy mode is not allowed.
```

**Diff**

```diff
diff --git a/doc/source/reference/c-api/array.rst b/doc/source/reference/c-api/array.rst
index 60037af8a..62f5648d9 100644
--- a/doc/source/reference/c-api/array.rst
+++ b/doc/source/reference/c-api/array.rst
@@ -633,7 +633,7 @@ From other objects
     borrowed reference to :c:data:`Py_NotImplemented` is returned.
 
 .. c:function:: PyObject* PyArray_FromArrayAttr( \
-        PyObject* op, PyArray_Descr* dtype, PyObject* context)
+        PyObject* op, PyArray_Descr* dtype, PyObject* context, int copy_mode)
 
     Return an ndarray object from a Python object that exposes the
     :obj:`~numpy.class.__array__` method. The :obj:`~numpy.class.__array__`
diff --git a/numpy/__init__.cython-30.pxd b/numpy/__init__.cython-30.pxd
index 42a46d0b8..4aa8117b5 100644
--- a/numpy/__init__.cython-30.pxd
+++ b/numpy/__init__.cython-30.pxd
@@ -623,7 +623,7 @@
     #object PyArray_FromArray (ndarray, dtype, int)
     object PyArray_FromInterface (object)
     object PyArray_FromStructInterface (object)
-    #object PyArray_FromArrayAttr (object, dtype, object)
+    #object PyArray_FromArrayAttr (object, dtype, object, int)
     #NPY_SCALARKIND PyArray_ScalarKind (int, ndarray*)
     int PyArray_CanCoerceScalar (int, int, NPY_SCALARKIND)
     object PyArray_NewFlagsObject (object)
diff --git a/numpy/__init__.pxd b/numpy/__init__.pxd
index 97f3da2e5..768bbe0bd 100644
--- a/numpy/__init__.pxd
+++ b/numpy/__init__.pxd
@@ -581,7 +581,7 @@
     #object PyArray_FromArray (ndarray, dtype, int)
     object PyArray_FromInterface (object)
     object PyArray_FromStructInterface (object)
-    #object PyArray_FromArrayAttr (object, dtype, object)
+    #object PyArray_FromArrayAttr (object, dtype, object, int)
     #NPY_SCALARKIND PyArray_ScalarKind (int, ndarray*)
     int PyArray_CanCoerceScalar (int, int, NPY_SCALARKIND)
     object PyArray_NewFlagsObject (object)
diff --git a/numpy/core/include/numpy/ndarrayobject.h b/numpy/core/include/numpy/ndarrayobject.h
index 2eb951486..81a3ff0b7 100644
--- a/numpy/core/include/numpy/ndarrayobject.h
+++ b/numpy/core/include/numpy/ndarrayobject.h
@@ -31,7 +31,7 @@ extern "C" {
 #define PyArray_HasArrayInterfaceType(op, type, context, out)                 \
         ((((out)=PyArray_FromStructInterface(op)) != Py_NotImplemented) ||    \
          (((out)=PyArray_FromInterface(op)) != Py_NotImplemented) ||          \
-         (((out)=PyArray_FromArrayAttr(op, type, context)) !=                 \
+         (((out)=PyArray_FromArrayAttr(op, type, context, int)) !=                 \
           Py_NotImplemented))
 
 #define PyArray_HasArrayInterface(op, out)                                    \
diff --git a/numpy/core/src/multiarray/array_coercion.c b/numpy/core/src/multiarray/array_coercion.c
index 847bdafc3..b2941a3ed 100644
--- a/numpy/core/src/multiarray/array_coercion.c
+++ b/numpy/core/src/multiarray/array_coercion.c
@@ -865,7 +865,8 @@ PyArray_DiscoverDTypeAndShape_Recursive(
         PyObject *obj, int curr_dims, int max_dims, PyArray_Descr**out_descr,
         npy_intp out_shape[NPY_MAXDIMS],
         coercion_cache_obj ***coercion_cache_tail_ptr,
-        PyArray_DTypeMeta *fixed_DType, enum _dtype_discovery_flags *flags)
+        PyArray_DTypeMeta *fixed_DType, enum _dtype_discovery_flags *flags,
+        int copy_mode)
 {
     PyArrayObject *arr = NULL;
     PyObject *seq;
@@ -923,7 +924,7 @@ PyArray_DiscoverDTypeAndShape_Recursive(
             requested_descr = *out_descr;
         }
         arr = (PyArrayObject *)_array_from_array_like(obj,
-                requested_descr, 0, NULL);
+                requested_descr, 0, NULL, copy_mode);
         if (arr == NULL) {
             return -1;
         }
@@ -1117,7 +1118,7 @@ PyArray_DiscoverDTypeAndShape_Recursive(
         max_dims = PyArray_DiscoverDTypeAndShape_Recursive(
                 objects[i], curr_dims + 1, max_dims,
                 out_descr, out_shape, coercion_cache_tail_ptr, fixed_DType,
-                flags);
+                flags, copy_mode);
 
         if (max_dims < 0) {
             return -1;
@@ -1169,7 +1170,7 @@ PyArray_DiscoverDTypeAndShape(
         npy_intp out_shape[NPY_MAXDIMS],
         coercion_cache_obj **coercion_cache,
         PyArray_DTypeMeta *fixed_DType, PyArray_Descr *requested_descr,
-        PyArray_Descr **out_descr)
+        PyArray_Descr **out_descr, int copy_mode)
 {
     coercion_cache_obj **coercion_cache_head = coercion_cache;
     *coercion_cache = NULL;
@@ -1214,7 +1215,7 @@ PyArray_DiscoverDTypeAndShape(
 
     int ndim = PyArray_DiscoverDTypeAndShape_Recursive(
             obj, 0, max_dims, out_descr, out_shape, &coercion_cache,
-            fixed_DType, &flags);
+            fixed_DType, &flags, copy_mode);
     if (ndim < 0) {
         goto fail;
     }
@@ -1499,7 +1500,7 @@ _discover_array_parameters(PyObject *NPY_UNUSED(self),
     int ndim = PyArray_DiscoverDTypeAndShape(
             obj, NPY_MAXDIMS, shape,
             &coercion_cache,
-            fixed_DType, fixed_descriptor, (PyArray_Descr **)&out_dtype);
+            fixed_DType, fixed_descriptor, (PyArray_Descr **)&out_dtype, 0);
     Py_XDECREF(fixed_DType);
     Py_XDECREF(fixed_descriptor);
     if (ndim < 0) {
diff --git a/numpy/core/src/multiarray/array_coercion.h b/numpy/core/src/multiarray/array_coercion.h
index db0e479fe..8f06d30fc 100644
--- a/numpy/core/src/multiarray/array_coercion.h
+++ b/numpy/core/src/multiarray/array_coercion.h
@@ -31,7 +31,7 @@ PyArray_DiscoverDTypeAndShape(
         npy_intp out_shape[NPY_MAXDIMS],
         coercion_cache_obj **coercion_cache,
         PyArray_DTypeMeta *fixed_DType, PyArray_Descr *requested_descr,
-        PyArray_Descr **out_descr);
+        PyArray_Descr **out_descr, int copy_mode);
 
 NPY_NO_EXPORT int
 PyArray_ExtractDTypeAndDescriptor(PyObject *dtype,
diff --git a/numpy/core/src/multiarray/arrayobject.c b/numpy/core/src/multiarray/arrayobject.c
index c8aaced4e..1b197d0f2 100644
--- a/numpy/core/src/multiarray/arrayobject.c
+++ b/numpy/core/src/multiarray/arrayobject.c
@@ -263,7 +263,7 @@ PyArray_CopyObject(PyArrayObject *dest, PyObject *src_object)
      */
     ndim = PyArray_DiscoverDTypeAndShape(src_object,
             PyArray_NDIM(dest), dims, &cache,
-            NPY_DTYPE(PyArray_DESCR(dest)), PyArray_DESCR(dest), &dtype);
+            NPY_DTYPE(PyArray_DESCR(dest)), PyArray_DESCR(dest), &dtype, 0);
     if (ndim < 0) {
         return -1;
     }
diff --git a/numpy/core/src/multiarray/common.c b/numpy/core/src/multiarray/common.c
index 82d34193d..aa95d285a 100644
--- a/numpy/core/src/multiarray/common.c
+++ b/numpy/core/src/multiarray/common.c
@@ -119,7 +119,7 @@ PyArray_DTypeFromObject(PyObject *obj, int maxdims, PyArray_Descr **out_dtype)
     int ndim;
 
     ndim = PyArray_DiscoverDTypeAndShape(
-            obj, maxdims, shape, &cache, NULL, NULL, out_dtype);
+            obj, maxdims, shape, &cache, NULL, NULL, out_dtype, 0);
     if (ndim < 0) {
         return -1;
     }
diff --git a/numpy/core/src/multiarray/ctors.c b/numpy/core/src/multiarray/ctors.c
index a8f41f582..a7acea70a 100644
--- a/numpy/core/src/multiarray/ctors.c
+++ b/numpy/core/src/multiarray/ctors.c
@@ -1291,7 +1291,8 @@ fail:
  */
 NPY_NO_EXPORT PyObject *
 _array_from_array_like(PyObject *op,
-        PyArray_Descr *requested_dtype, npy_bool writeable, PyObject *context) {
+        PyArray_Descr *requested_dtype, npy_bool writeable, PyObject *context,
+        int copy_mode) {
     PyObject* tmp;
 
     /*
@@ -1347,7 +1348,7 @@ _array_from_array_like(PyObject *op,
      *      this should be changed!
      */
     if (!writeable && tmp == Py_NotImplemented) {
-        tmp = PyArray_FromArrayAttr(op, requested_dtype, context);
+        tmp = PyArray_FromArrayAttr(op, requested_dtype, context, copy_mode);
         if (tmp == NULL) {
             return NULL;
         }
@@ -1447,7 +1448,7 @@ setArrayFromSequence(PyArrayObject *a, PyObject *s,
     }
 
     /* Try __array__ before using s as a sequence */
-    PyObject *tmp = _array_from_array_like(s, NULL, 0, NULL);
+    PyObject *tmp = _array_from_array_like(s, NULL, 0, NULL, 0);
     if (tmp == NULL) {
         goto fail;
     }
@@ -1575,7 +1576,8 @@ PyArray_FromAny(PyObject *op, PyArray_Descr *newtype, int min_depth,
     Py_XDECREF(newtype);
 
     ndim = PyArray_DiscoverDTypeAndShape(op,
-            NPY_MAXDIMS, dims, &cache, fixed_DType, fixed_descriptor, &dtype);
+            NPY_MAXDIMS, dims, &cache, fixed_DType, fixed_descriptor, &dtype,
+            flags & _NPY_ARRAY_ENSURENOCOPY);
 
     Py_XDECREF(fixed_descriptor);
     Py_XDECREF(fixed_DType);
@@ -2459,7 +2461,7 @@ PyArray_FromInterface(PyObject *origin)
 /*NUMPY_API
  */
 NPY_NO_EXPORT PyObject *
-PyArray_FromArrayAttr(PyObject *op, PyArray_Descr *typecode, PyObject *context)
+PyArray_FromArrayAttr(PyObject *op, PyArray_Descr *typecode, PyObject *context, int copy_mode)
 {
     PyObject *new;
     PyObject *array_meth;
@@ -2493,6 +2495,10 @@ PyArray_FromArrayAttr(PyObject *op, PyArray_Descr *typecode, PyObject *context)
         Py_DECREF(array_meth);
         return Py_NotImplemented;
     }
+    if( copy_mode ) {
+        PyErr_SetString(PyExc_ValueError, "Calling __array__ in never copy mode is not allowed.");
+        return NULL;
+    }
     if (typecode == NULL) {
         new = PyObject_CallFunction(array_meth, NULL);
     }
diff --git a/numpy/core/src/multiarray/ctors.h b/numpy/core/src/multiarray/ctors.h
index e59e86e8b..1dae26173 100644
--- a/numpy/core/src/multiarray/ctors.h
+++ b/numpy/core/src/multiarray/ctors.h
@@ -32,7 +32,7 @@ PyArray_New(
 
 NPY_NO_EXPORT PyObject *
 _array_from_array_like(PyObject *op,
-        PyArray_Descr *requested_dtype, npy_bool writeable, PyObject *context);
+        PyArray_Descr *requested_dtype, npy_bool writeable, PyObject *context, int copy_mode);
 
 NPY_NO_EXPORT PyObject *
 PyArray_FromAny(PyObject *op, PyArray_Descr *newtype, int min_depth,
@@ -53,7 +53,7 @@ PyArray_FromInterface(PyObject *input);
 
 NPY_NO_EXPORT PyObject *
 PyArray_FromArrayAttr(PyObject *op, PyArray_Descr *typecode,
-                      PyObject *context);
+                      PyObject *context, int copy_mode);
 
 NPY_NO_EXPORT PyObject *
 PyArray_EnsureArray(PyObject *op);

```

Let me know if it looks fine. I am not sure if it modifies any public APIs.

Thanks. cc: @mattip @seberg

Reply:
Yes, something like that should be right.

Reply:
I think a better place to check this is [just before the call to `PyArray_FromArrayAttr`](https://github.com/numpy/numpy/pull/19173/files#diff-6897a5d4c65432ae60cb14abe2e802943033f612500418a3766a64f4bbbb0480R1350). There is already a check for `writeable` there, and by doing it there you do not have to change public API.

Reply:
Since, @mattip suggested that we shouldn't modify the public `PyArray_FromArrayAttr` function. I will add the check for never copy flag one step above in the call ladder for `__array__` method.

Reply:
gahh, confusion. The deleted comment was more correct, please do not change `PyArray_FromArrayAttr` rather just above its call in `_array_from_array_like`

Reply:
Ah, I had not noticed that, yeah I guess `PyArray_FromArrayAttr` is public, but all others are private.